modules/new_serverpod_auth/serverpod_auth_email_account/serverpod_auth_email_account_server/lib/src/business/email_accounts.dart

```diff
@@ -0,0 +1,314 @@
+    );
+
+    account ??=
+        await _importExistingUser(session, email: email, password: password);
+
+    if (account == null) {
+      throw EmailAccountLoginException(
+        reason: EmailAccountLoginFailureReason.invalidCredentials,
+      );
+    }
+
+    if (await _hasTooManyFailedSignIns(session, email)) {
+      throw EmailAccountLoginException(
+        reason: EmailAccountLoginFailureReason.tooManyAttempts,
+      );
+    }
+
+    if (!PasswordHash.validateHash(
+      email: email,
+      password: password,
+      hash: account.passwordHash,
+    )) {
+      await _logFailedSignIn(session, email);
+
+      throw EmailAccountLoginException(
+        reason: EmailAccountLoginFailureReason.invalidCredentials,
+      );
+    }
+
+    return account.userId;
+  }
+
+  /// Returns the registration process ID.
+  ///
+  /// The caller may store additional information attached to this ID,
+  /// which will be returned from [verify] later on.
+  static Future<UuidValue> requestAccount(
+    final Session session, {
+    required String email,
+    required String password,
+  }) async {
+    email = email.trim().toLowerCase();
+    password = password.trim();
+
+    final existingAccountCount = await EmailAccount.db
+        .count(session, where: (final t) => t.email.equals(email));
+    if (existingAccountCount > 0) {
+      throw Exception('Email already registered');
+    }
+
+    final verificationCode = Random.secure().nextString(length: 20);
+
+    final request = await EmailAccountRequest.db.insertRow(
+      session,
+      EmailAccountRequest(
+        email: email,
+        passwordHash: password,
+        created: DateTime.now(),
+        expiration: DateTime.now().add(
+          EmailAccountConfig.current.registrationVerificationCodeLifetime,
+        ),
+        verificationCode: verificationCode,
+      ),
+    );
+
+    EmailAccountConfig.current.sendRegistrationVerificationMail?.call(
+      email: email,
+      verificationToken: verificationCode,
+    );
+
+    return request.id!;
+  }
+
+  /// Returns the account request creation ID if the request is valid.
+  ///
+  /// If this returns a value, this means `createAccount` will succeed.
+  static Future<(UuidValue accountRequestId, String email)?>
+      verifyAccountRequest(
+    final Session session, {
+    required final String verificationCode,
+  }) async {
+    final request = (await EmailAccountRequest.db.find(session,
+            where: (final t) => t.verificationCode.equals(verificationCode)))
+        .where((final r) => r.expiration.isAfter(DateTime.now()))
+        .firstOrNull;
+
+    if (request == null) {
+      return null;
+    }
+
+    return (request.id!, request.email);
+  }
+
+  /// Finalize the email authentication creation.
+  ///
+  /// Returns the `ID` of the new email authentication, and the email address used during registration.
+  static Future<(UuidValue emailAuthenticationId, String emailAddress)>
```

Comment:
I wonder if named records would be preferred here.

Because there is the `emailAuthenticationId` and `emailAuthenticationRequestId`, and that is not obvious from the caller without reading the docs (plus the email being returned (instead of the whole object, which is not that useful outside), as the caller might want to create a profile).

Reply:
switched to named records for clarity